### PR TITLE
Rewrite WebSocket class

### PR DESCRIPTION
**Pull request by alexandrebouthinon**
## What does this PR do ?
The WebSocket protocol class has been rewritten. 
Before this PR a new SDK instance was create like this:
```cs
KuzzleSdk.Kuzzle kuzzle = new KuzzleSdk.Kuzzle(new WebSocket("kuzzle"));
await kuzzle.ConnectAsync();
```
now this can be done like this:
```cs
KuzzleSdk.Kuzzle kuzzle = new KuzzleSdk.Kuzzle(new WebSocket(new Uri("ws://kuzzle:7512")));
kuzzle.ConnectAsync(CancellationToken.None).Wait();
```

### How should this be manually tested?

Run unit tests

### Other changes

* Add unit tests.
* Fix documentation tests template.

### CODE REVIEW

**Review by Shiranuit — APPROVED** [review]

File: .ci/doc/templates/default.tpl.cs [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
-using KuzzleSdk.API;
 using KuzzleSdk.Protocol;
-using System.Threading.Tasks;
 
-KuzzleSdk.Kuzzle kuzzle = new KuzzleSdk.Kuzzle(new WebSocket("kuzzle"));
+WebSocket socket = new WebSocket(new Uri("ws://kuzzle:7512"));
+KuzzleSdk.Kuzzle kuzzle = new KuzzleSdk.Kuzzle(socket);
 
-await kuzzle.ConnectAsync();
+kuzzle.ConnectAsync(CancellationToken.None).Wait();
```
Comment (benoitvidis) [inline .ci/doc/templates/default.tpl.cs:11]: Is the `<method>.Wait()` syntax an alternative to `await`? And if so, is there a reason to prefer it over `await`?

File: Kuzzle/Protocol/WebSocket.cs [hunk 2]
```diff
@@ -7,151 +8,167 @@
...
-    public int Port {
-      get { return port ?? 7512; }
-      set { port = value; }
-    }
-
-    /// <summary>
-    /// Timeout (in milliseconds) after which a connection attempt aborts.
-    /// </summary>
-    public int ConnectTimeout {
-      get { return connectTimeout ?? 30000; }
```
Comment (benoitvidis) [inline Kuzzle/Protocol/WebSocket.cs:37]: Shouldn't we keep a default timeout?

**Review by benoitvidis — CHANGES_REQUESTED** [review]
I agree that passing a `Uri` instance to the `WebSocket` protocol is simpler than splitting it between a `host` string and a set of options but this means we will end-up with totally different protocol constructors per language, which, IMO, is not a problem per-se.

The use of a mandatory `CancellationToken` on the `ConnectAsync` method is more problematic to me.
As the end-user is unable to act on the token, the only possible action for him is to generate a self-expiring token, which was previously handled within the default constructor options.

I prefer this last option. Additionally, we could extend it to allow the user to declare his own ws client buffer sizes for instance.

So, to sum up, I would remove the `CancellationToken` parameter from the `ConnectAsync` methods and generate one based on the options given to the protocol constructor instead.

File: .ci/doc/templates/default.tpl.cs (hunk 1 above)
Comment (alexandrebouthinon) [reply .ci/doc/templates/default.tpl.cs:11]: .Wait() is a Thread method to wait for the thread termination. I used it here because main function cannot be async.

Comment (Aschen) [reply .ci/doc/templates/default.tpl.cs:11]: Main function does not need to be async, this script is run with `dotnet script` which allow `await` keyword in the script root without encapsulating the code in a main async function

File: Kuzzle/Protocol/WebSocket.cs (hunk 2 above)
Comment (alexandrebouthinon) [reply Kuzzle/Protocol/WebSocket.cs]: > I agree that passing a `Uri` instance to the `WebSocket` protocol is simpler than splitting it between a `host` string and a set of options but this means we will end-up with totally different protocol constructors per language, which, IMO, is not a problem per-se.
> 
> The use of a mandatory `CancellationToken` on the `ConnectAsync` method is more problematic to me.
> As the end-user is unable to act on the token, the only possible action for him is to generate a self-expiring token, which was previously handled within the default constructor options.
> 
> I prefer this last option. Additionally, we could extend it to allow the user to declare his own ws client buffer sizes for instance.
> 
> So, to sum up, I would remove the `CancellationToken` parameter from the `ConnectAsync` methods and generate one based on the options given to the protocol constructor instead.

After a discuss with @scottinet, this is ugly but a C# idiomatic way to pass optional params (in fact, nothing is optional, all is explicitly given)

**Comment by alexandrebouthinon** [comment]
@benoitvidis:
> I agree that passing a `Uri` instance to the `WebSocket` protocol is simpler than splitting it between a `host` string and a set of options but this means we will end-up with totally different protocol constructors per language, which, IMO, is not a problem per-se.
> 
> The use of a mandatory `CancellationToken` on the `ConnectAsync` method is more problematic to me.
> As the end-user is unable to act on the token, the only possible action for him is to generate a self-expiring token, which was previously handled within the default constructor options.
> 
> I prefer this last option. Additionally, we could extend it to allow the user to declare his own ws client buffer sizes for instance.
> 
> So, to sum up, I would remove the `CancellationToken` parameter from the `ConnectAsync` methods and generate one based on the options given to the protocol constructor instead.
> I agree that passing a `Uri` instance to the `WebSocket` protocol is simpler than splitting it between a `host` string and a set of options but this means we will end-up with totally different protocol constructors per language, which, IMO, is not a problem per-se.
> 
> The use of a mandatory `CancellationToken` on the `ConnectAsync` method is more problematic to me.
> As the end-user is unable to act on the token, the only possible action for him is to generate a self-expiring token, which was previously handled within the default constructor options.
> 
> I prefer this last option. Additionally, we could extend it to allow the user to declare his own ws client buffer sizes for instance.
> 
> So, to sum up, I would remove the `CancellationToken` parameter from the `ConnectAsync` methods and generate one based on the options given to the protocol constructor instead.

After a discuss with @scottinet, this is ugly but it seems to be a C# idiomatic way to pass optional params (in fact, nothing is optional, all is explicitly given). In addition, this optimization of the performance of the Websocket class is expected and it is an alpha version. We can change the behavior of this function later. :slightly_smiling_face:

File: Kuzzle/Protocol/WebSocket.cs (hunk 2 above)
Comment (alexandrebouthinon) [reply Kuzzle/Protocol/WebSocket.cs:37]: see my comment below :slightly_smiling_face:

File: .ci/doc/templates/default.tpl.cs (hunk 1 above)
Comment (alexandrebouthinon) [reply .ci/doc/templates/default.tpl.cs:11]: Since it is an example of code, should we assume that it will be used in a main function?

**Comment by benoitvidis** [comment]
@alexandrebouthinon , @scottinet , I am sorry, I do not understand the response.

Do you mean c# discourages the use of default parameters? The [only reference I found](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1026-default-parameters-should-not-be-used?view=vs-2019) only encourages using method overloads over default parameters.
To me, whether the connection options are given to the constructor or the connect method are totally driven by design, not by the language. 

To get the expected default timeout behavior with this proposed implementation, the developer must write this kind of code:

```csharp
WebSocket socket = new KuzzleSdk.Protocol.WebSocket(new Uri("wss://localhost:7512"));
var kuzzle = new KuzzleSdk.Kuzzle(socket);

var source = new CancellationTokenSource(30000);
await kuzzle.ConnectAsync(source.Token);
```
This creates yet some other extra verbosity and make It quite easy to forget / skip the `CancellationTokenSource` creation imo.

Further, what about the auto-reconnect? Which `CancellationToken` is used then?

**About other options**

`ClientWebSocket` supports several other options, including the ability to set some additional http headers or to set the size of the request/answer buffers.

I believe we should give the ability to the developer to access/modify these properties.

**About signature and alpha versions**

My understanding is that this version is meant to be distributed and used quite soon by a sample of users. Any further breaking change will imply some rework for them.
I believe we should try as hard as we can to define some solid signatures from the start.

**Comment by softlair38** [comment]
@benoitvidis  : I am agree that CancellationToken must be manage by the library and not pass in the constructor. When call Close, it must call Cancel() in cancellationToken.

Default timeout can be 30 but it need to be modified, like the buffer size.

Methods with await, could terminate with .ConfigureAwait(false);

You can use ConcurrentQueue instead of BlockingCollection to keep the order of message

**Comment by Shiranuit** [comment]
@softlair38
> You can use ConcurrentQueue instead of BlockingCollection to keep the order of message

The default collection type for `BlockingCollection<T>` is `ConcurrentQueue<T>`, so no need to change it.

**Comment by softlair38** [comment]
@Shiranuit yes, you are right, I did not use this class until now, thank you

**Comment by scottinet** [comment]
@benoitvidis > I'm afraid the reasons I gave to Alexandre were a bit lost in translation.
What I said was that it was idiomatic to C#, not because it's ugly or because C# encourages explicit paramters (it does not), but because C# API makes sure that it's obvious that new threads are created when invoking such methods.
It's a lot safer for maintenability reason to pass an explicit `CancellationToken.None` option when explicitly forfeiting any control to the created thread, instead of having default values that may hide the fact that a new thread is started.

And I think that we should do the same.

**Review by benoitvidis — APPROVED** [review]
